cmd/thanos/sidecar.go

```diff
@@ -286,19 +286,23 @@ func runSidecar(
 			info.WithMetricMetadataInfoFunc(),
 		)
 
-		storeServer := store.NewLimitedStoreServer(store.NewInstrumentedStoreServer(reg, promStore), reg, conf.storeRateLimits)
-		s := grpcserver.New(logger, reg, tracer, grpcLogOpts, tagOpts, comp, grpcProbe,
-			grpcserver.WithServer(store.RegisterStoreServer(storeServer, logger)),
-			grpcserver.WithServer(rules.RegisterRulesServer(rules.NewPrometheus(conf.prometheus.url, c, m.Labels))),
+		opts := []grpcserver.Option{
 			grpcserver.WithServer(targets.RegisterTargetsServer(targets.NewPrometheus(conf.prometheus.url, c, m.Labels))),
-			grpcserver.WithServer(meta.RegisterMetadataServer(meta.NewPrometheus(conf.prometheus.url, c))),
-			grpcserver.WithServer(exemplars.RegisterExemplarsServer(exemplarSrv)),
-			grpcserver.WithServer(info.RegisterInfoServer(infoSrv)),
 			grpcserver.WithListen(conf.grpc.bindAddress),
 			grpcserver.WithGracePeriod(conf.grpc.gracePeriod),
 			grpcserver.WithMaxConnAge(conf.grpc.maxConnectionAge),
 			grpcserver.WithTLSConfig(tlsCfg),
-		)
+		}
+		if !promAgentModeEnabled {
+			storeServer := store.NewLimitedStoreServer(store.NewInstrumentedStoreServer(reg, promStore), reg, conf.storeRateLimits)
+			opts = append(opts, grpcserver.WithServer(store.RegisterStoreServer(storeServer, logger)),
+				grpcserver.WithServer(rules.RegisterRulesServer(rules.NewPrometheus(conf.prometheus.url, c, m.Labels))),
+				grpcserver.WithServer(meta.RegisterMetadataServer(meta.NewPrometheus(conf.prometheus.url, c))),
+				grpcserver.WithServer(exemplars.RegisterExemplarsServer(exemplarSrv)),
+				grpcserver.WithServer(info.RegisterInfoServer(infoSrv)),
+			)
+		}
```

Comment:
So I think, idea is to disable only the storeAPI in agent mode, but not the others?
So `meta`, `info`, `rules` should probably be with `targets` server. And we conditionally enable `store` and `exemplars`.

Also, I think the `info` server probably deserves some changes, so that queries don't get fanned out to the Sidecar in agent mode?